### PR TITLE
third_party dependencies

### PR DESCRIPTION
**Pull request by pca006132**
Reincarnation of #402, this uses system gtest if possible and automatically download gtest from github otherwise (e.g. for windows). It will also download tbb and compile from source if the user specified `MANIFOLD_PAR=TBB` and CMake is unable to find the installation.

I also added optional dependency to [Tracy profiler](https://github.com/wolfpld/tracy), which is a handy graphical profiler.

### CODE REVIEW

File: src/manifold/src/manifold.cpp [hunk 1]
```diff
@@ -260,7 +260,10 @@ MeshGL Manifold::GetMeshGL(glm::ivec3 normalIdx) const {
       out.triVerts[3 * tri + i] = impl.halfedge_[3 * oldTri + i].startVert;
 
     if (meshID != lastID) {
-      addRun(out, runNormalTransform, tri, meshIDtransform.at(meshID));
+      Impl::Relation rel;
+      auto it = meshIDtransform.find(meshID);
+      if (it != meshIDtransform.end()) rel = it->second;
```
Comment (elalish) [inline src/manifold/src/manifold.cpp:265]: Is there a case where this is expected, or should this be an assert?

File: src/utilities/CMakeLists.txt [hunk 2]
```diff
@@ -29,11 +29,32 @@ if(MANIFOLD_PAR STREQUAL "OMP")
...
 elseif(MANIFOLD_PAR STREQUAL "TBB")
-    find_package(PkgConfig REQUIRED)
-    pkg_check_modules(TBB REQUIRED tbb)
-    target_include_directories(${PROJECT_NAME} PUBLIC ${TBB_INCLUDE_DIRS})
+    find_package(PkgConfig)
+    if (PKG_CONFIG_FOUND)
+        pkg_check_modules(TBB tbb)
+    endif()
+    if(NOT TBB_FOUND)
+        message(STATUS "tbb not found, downloading from source")
```
Comment (elalish) [inline src/utilities/CMakeLists.txt:37]: 💯

File: src/utilities/include/public.h [hunk 3]
```diff
@@ -19,17 +19,14 @@
...
 #include <memory>
+#include <sstream>
 #include <unordered_map>
 #include <vector>
 
-#ifdef MANIFOLD_DEBUG
-#include <iomanip>
-#include <iostream>
-#include <sstream>
-#endif
```
Comment (elalish) [inline src/utilities/include/public.h:31]: Why this change? Part of my purpose here was to keep the WASM build size down since including things like `std::string` are notoriously large.

File: test/CMakeLists.txt [hunk 4]
```diff
@@ -24,9 +22,45 @@ if(MANIFOLD_CBIND)
...
 endif()
 
+set(gtest_force_shared_crt ON CACHE BOOL "" FORCE)
+# Prevent installation of GTest with your project
+set(INSTALL_GTEST OFF CACHE BOOL "" FORCE)
+set(INSTALL_GMOCK OFF CACHE BOOL "" FORCE)
+
+find_package(GTest 1.12)
+if((NOT GTest_FOUND) OR NOT(GTest_VERSION GREATER_EQUAL 1.12))
+  message(STATUS "GTest not found, downloading from source")
```
Comment (elalish) [inline test/CMakeLists.txt:32]: @kintel - what do you think of this approach? Would this make consuming our library easier for the likes of OpenSCAD?

**Review by elalish — COMMENTED** [review]
Awesome, this looks great! Can you give any example of what you can do with Tracy? I'm not familiar.

File: src/manifold/src/manifold.cpp (hunk 1 above)
Comment (pca006132) [reply src/manifold/src/manifold.cpp:265]: I am not entirely sure, it seems that the our python test cases would somehow trigger this.

File: src/utilities/include/public.h (hunk 3 above)
Comment (pca006132) [reply src/utilities/include/public.h:31]: Oh I forgot to revert this change, I was trying to debug the ubuntu failure when using system provided gtest.

**Comment by pca006132** [comment]
I am not familiar with Tracy either :) We should be able to add more client markups to make the profiler easier to use.
For now, we can see the CPU utilization and roughly what functions are running.

image
image

**Comment by pca006132** [comment]
Sadly it seems that some features are not available on OSX.
image

File: src/manifold/src/manifold.cpp (hunk 1 above)
Comment (elalish) [reply src/manifold/src/manifold.cpp:265]: That seems worth investigating. I think it implies we have marked some faces as coming from an ID for which there is no transform. Those things should be 1:1. We definitely don't have much testing of this in the Python bindings, so we could have a bug there.

Comment (pca006132) [reply src/manifold/src/manifold.cpp:265]: indeed, I think I saw this happening previously as well

Comment (pca006132) [reply src/manifold/src/manifold.cpp:265]: maybe we can make an issue for this and merge this first?

Comment (elalish) [reply src/manifold/src/manifold.cpp:265]: Sounds good.

**Review by elalish — APPROVED** [review]
Thanks!

File: test/CMakeLists.txt (hunk 4 above)
Comment (kintel) [reply test/CMakeLists.txt:32]: Looks reasonable. It does make the build system depend on having access to github.com, which is a known issue from some countries, as well as for people with limited connectivity, but that's a separate challenge. For OpenSCAD we'll likely try to hit the find_package call, so this should work well for us!

Comment (pca006132) [reply test/CMakeLists.txt:32]: Yes it depends on access to github by default, but if for some reason that is not possible and `find_package` still cannot find the package, you can specify the directory by setting `-DFETCHCONTENT_SOURCE_DIR_GOOGLETEST`. I tried using this for the nix package (it blocks internet connection when building) and it works.

Comment (kintel) [reply test/CMakeLists.txt:32]: Awesom! Not familiar with FetchContent..

Comment (elalish) [reply test/CMakeLists.txt:32]: Nice - would be good to add that trick to the README somewhere.
